File: back/src/main/java/com/tdevilleduc/urthehero/back/controller/PageController.java
package com.tdevilleduc.urthehero.back.controller;

import com.tdevilleduc.urthehero.back.exceptions.PageInternalErrorException;
import com.tdevilleduc.urthehero.back.model.Page;
import com.tdevilleduc.urthehero.back.model.PageDTO;
import com.tdevilleduc.urthehero.back.model.Story;
import com.tdevilleduc.urthehero.back.service.IPageService;
import com.tdevilleduc.urthehero.back.service.IStoryService;
import io.swagger.annotations.Api;
import io.swagger.annotations.ApiOperation;
import lombok.extern.slf4j.Slf4j;
import org.slf4j.helpers.MessageFormatter;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.util.Assert;
import org.springframework.web.bind.annotation.*;

import javax.servlet.http.HttpServletRequest;
import javax.validation.constraints.NotNull;
import java.util.Optional;
import java.util.concurrent.Callable;

@Slf4j
@Api(value = "Page", tags = { "Page Controller" } )
@RestController
@RequestMapping("/api/page")
class PageController {

    private final IStoryService storyService;
    private final IPageService pageService;

    public PageController(IStoryService storyService, IPageService pageService) {
        this.storyService = storyService;
        this.pageService = pageService;
    }

    @GetMapping(value = "/{pageId}", produces = MediaType.APPLICATION_JSON_VALUE)
    @ResponseStatus(HttpStatus.OK)
    @ApiOperation(
            value = "${swagger.controller.page.get-by-id.value}",
            notes = "${swagger.controller.page.get-by-id.notes}")
    public @ResponseBody Callable<ResponseEntity<Page>> getPageById(HttpServletRequest request,
                                                                    @PathVariable Integer pageId) {
        return () -> {
            if (log.isInfoEnabled()) {
                log.info("call: {}", request.getRequestURI());
            }
            Optional<Page> optional = pageService.findById(pageId);
            return optional
                    .map(ResponseEntity::ok)
                    .orElseGet(() -> ResponseEntity.notFound().build());
        };
    }

    @GetMapping(value = "/story/{storyId}", produces = MediaType.APPLICATION_JSON_VALUE)
    @ResponseStatus(HttpStatus.OK)
    @ApiOperation(
            value = "${swagger.controller.page.get-first-by-story-id.value}",
            notes = "${swagger.controller.page.get-first-by-story-id.notes}")
    public @ResponseBody Callable<ResponseEntity<Page>> getFirstPageByStoryId(HttpServletRequest request,
                                                                              @PathVariable int storyId) {
        return () -> {
            if (log.isInfoEnabled()) {
                log.info("call: {}", request.getRequestURI());
            }
            Optional<Story> optional = storyService.findById(storyId);
            return optional
                    .map(Story::getFirstPageId)
                    .map(pageService::findById)
                    .map(Optional::get)
                    .map(ResponseEntity::ok)
                    .orElseGet(() -> ResponseEntity.notFound().build());
        };
    }

    @PutMapping
    public @ResponseBody Callable<ResponseEntity<PageDTO>> createPage(HttpServletRequest request,
                           @RequestBody @NotNull PageDTO page) {
        return () -> {
            if (log.isInfoEnabled()) {
                log.info("call: {}", request.getRequestURI());
            }
            if (page.getId() != null && pageService.exists(page.getId())) {
                throw new PageInternalErrorException(MessageFormatter.format("Une page avec l'identifiant {} existe déjà. Elle ne peut être créée", page.getId()).getMessage());
            }
            return ResponseEntity.ok(pageService.createOrUpdate(page));
        };
    }

    @PostMapping
    public @ResponseBody Callable<ResponseEntity<PageDTO>> updatePage(HttpServletRequest request,
                           @RequestBody @NotNull PageDTO page) {
        return () -> {
            if (log.isInfoEnabled()) {
                log.info("call: {}", request.getRequestURI());
            }
            Assert.notNull(page.getId(), () -> {
                throw new PageInternalErrorException("L'identifiant de la page passée en paramètre ne peut pas être null");
            });
            return ResponseEntity.ok(pageService.createOrUpdate(page));
        };
    }

    @DeleteMapping(value = "/{pageId}")
    public @ResponseBody void deletePage(HttpServletRequest request,
                           @PathVariable @NotNull Integer pageId) {
        if (log.isInfoEnabled()) {
            log.info("call: {}", request.getRequestURI());
        }
        pageService.delete(pageId);
    }
}

File: back/src/main/java/com/tdevilleduc/urthehero/back/controller/PageController.kt
package com.tdevilleduc.urthehero.back.controller

import com.tdevilleduc.urthehero.back.constant.ApplicationConstants
import com.tdevilleduc.urthehero.back.exceptions.PageInternalErrorException
import com.tdevilleduc.urthehero.back.model.Page
import com.tdevilleduc.urthehero.back.model.PageDTO
import com.tdevilleduc.urthehero.back.service.IPageService
import com.tdevilleduc.urthehero.back.service.IStoryService
import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.security.SecurityRequirement
import io.swagger.v3.oas.annotations.tags.Tag
import org.slf4j.helpers.MessageFormatter
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.util.Assert
import org.springframework.web.bind.annotation.*
import java.util.concurrent.Callable

@Tag(name = "Page", description = "Page Controller")
@RestController
@RequestMapping("/api/page")
@SecurityRequirement(name = "bearerAuth")
internal class PageController(private val storyService: IStoryService, private val pageService: IPageService) {

    @GetMapping(value = ["/{pageId}"], produces = [MediaType.APPLICATION_JSON_VALUE])
    @ResponseStatus(HttpStatus.OK)
    @Operation(summary = "\${swagger.controller.page.get-by-id.value}", description = "\${swagger.controller.page.get-by-id.notes}")
    @ResponseBody
    fun getPageById(@PathVariable pageId: Int): Callable<ResponseEntity<Page>> = Callable {
        ResponseEntity.ok(pageService.findById(pageId))
    }

    @GetMapping(value = ["/story/{storyId}"], produces = [MediaType.APPLICATION_JSON_VALUE])
    @ResponseStatus(HttpStatus.OK)
    @Operation(summary = "\${swagger.controller.page.get-first-by-story-id.value}", description = "\${swagger.controller.page.get-first-by-story-id.notes}")
    @ResponseBody
    fun getFirstPageByStoryId(@PathVariable storyId: Int): Callable<ResponseEntity<Page>> = Callable {
        val story = storyService.findById(storyId)
        val firstPage = pageService.findById(story.firstPageId!!)
        ResponseEntity.ok(firstPage)
    }

    @PutMapping
    @ResponseBody
    fun createPage(@RequestBody page: PageDTO): Callable<ResponseEntity<PageDTO>> = Callable {
        if (pageService.exists(page.id)) {
            throw PageInternalErrorException(MessageFormatter.format(ApplicationConstants.ERROR_MESSAGE_PAGE_PAGEID_ALREADY_EXISTS, page.id).message)
        }
        ResponseEntity.ok(pageService.createOrUpdate(page))
    }

    @PostMapping
    @ResponseBody
    fun updatePage(@RequestBody pageDto: PageDTO): Callable<ResponseEntity<PageDTO>> = Callable {
        Assert.notNull(pageDto.id) { throw PageInternalErrorException(ApplicationConstants.ERROR_MESSAGE_PAGE_PAGEID_CANNOT_BE_NULL) }
        if (pageService.notExists(pageDto.id)) {
            throw PageInternalErrorException(MessageFormatter.format(ApplicationConstants.ERROR_MESSAGE_PAGE_DOESNOT_EXIST, pageDto.id).message)
        }
        ResponseEntity.ok(pageService.createOrUpdate(pageDto))
    }

    @DeleteMapping(value = ["/{pageId}"])
    @ResponseBody
    fun deletePage(@PathVariable pageId: Int) = Callable {
        pageService.delete(pageId)
    }

}